### PR TITLE
Issue #1496: BC loader/overfit smoke (cheap-lane worker)

### PR DESCRIPTION
**Pull request by ll7**
## Summary

Issue #1496 (benchmark oracle imitation warm-start) is gated on SLURM/GPU compute and a larger balanced collection for the full BC-warm-start-vs-RL comparison. The maintainer disposition after PR #5914 resolved the dataset-materialization blocker was explicit about the next smallest CPU-achievable step:

> "This is enough to validate loading, schema compatibility, a tiny overfit, and end-to-end BC smoke execution. It is not a credible dataset for the predeclared BC-warm-start-versus-RL sample-efficiency or final-performance comparison."

This PR delivers exactly that bounded **BC loader/overfit smoke**. It is **NEW capability** — it actually loads the `expert_traj_v1` NPZ, enforces the split/leakage contract, and runs a real BC training step — and does **not** duplicate the prior readiness/preflight PRs (#3767/#4002/#4091/#4106/#4111/#4119/#4127/#4133/#4510/#5914), which only checked fail-closed contracts or registered pointers.

### What changed

- `robot_sf/training/oracle_imitation_bc_smoke.py` (new):
  - `load_expert_trajectory_dataset` — loads and structurally validates the `expert_traj_v1.npz` NPZ schema (required arrays, episode/step consistency).
  - `validate_split_leakage_contract` — enforces the issue #1496 acceptance gate "Split/leakage validation passes before training starts": every required split present, no episode id shared across splits, train seeds disjoint from validation/evaluation, evaluation held out from training. Fails closed on any violation.
  - `flatten_observation_action_pairs` — flattens the training split's structured per-step observations + actions into `(features, targets)` tensors (skips the large occupancy grid; compact state vector is correct for an overfit probe).
  - `run_bc_overfit_smoke` — tiny MLP (CPU), MSE on actions, records initial→final loss, **asserts strict loss reduction** (fail-closed if the loader/loop is broken), writes checkpoint + metrics + smoke manifest.
- `scripts/validation/run_oracle_imitation_bc_smoke.py` (new): CLI resolving the dataset directly (`--dataset-path`) or through the durable registry (`--registry` + `--private-artifact-root`, full trace-contract + checksum verified, fail-closed if absent).
- `tests/training/test_oracle_imitation_bc_smoke.py` (new): schema-faithful synthetic NPZ fixture; asserts schema validation, three split/leakage fail-closed paths, flatten-keeps-only-train, overfit loss-decrease + artifact emission, and zero-epoch fail-closed.

### Claim boundary (honest)

A checkpoint from this smoke is **smoke evidence** — it proves the loader, the split/leakage gate, and a BC training step execute end to end and that the tiny training split is *memorable*. It is **NOT** benchmark evidence, navigation-quality evidence, a warm-start-quality result, or a paper/dissertation claim. Manifests carry explicit `smoke_not_quality`, `training_performed`, and `not_full_warm_start_comparison` flags. The full BC-warm-start-vs-RL comparison remains gated on a larger balanced collection and SLURM/GPU compute.

The real 363 MB registered dataset (`expert_traj_v1.npz`, sha256 `433c797f…`) is **not present on this worker host**, so the smoke here is validated against a schema-faithful synthetic NPZ fixture that reproduces the exact job 13520 arrays and observation contract. Running against the real private artifact requires the authorized private artifact root (the CLI fails closed otherwise — verified below).

### Validation (CPU only)

```
uv run ruff check <3 files>          # All checks passed!
uv run ruff format --check <3 files> # 3 files already formatted
uv run pytest tests/training/test_oracle_imitation_bc_smoke.py -v
# 9 passed in 4.03s
uv run pytest tests/training/test_oracle_imitation_warm_start_readiness.py \
                tests/training/test_oracle_imitation_bc_smoke.py \
                tests/training/test_issue_1496_job_13520_registration.py -q
# 35 passed (no regression in adjacent #1496/readiness surfaces)
```

Overfit smoke on a 3-split schema-faithful fixture (CLI `--json`): `loss 0.6646 → 1.7e-10` (reduction `0.6646`, 22 train steps, splits `{train:2, validation:1, evaluation:1}`).

Fail-closed checks verified:
- CLI `--registry` without `--private-artifact-root` → `failed` ("requires --private-artifact-root").
- CLI `--registry` with a root lacking the full trace set → `failed` (registry trace contract not training-ready).
- A train-only NPZ missing validation/evaluation splits → `failed` ("missing required split(s)").

### Successor statement

This slice does **not** duplicate prior readiness/preflight PRs: those established fail-closed contracts and registered pointers, while this PR delivers the first actual BC data loader + training step that consumes the `expert_traj_v1` artifact. The remaining acceptance criteria stay open: train a warm-start policy on a **larger balanced collection** (multiple nontrivial trajectories per scenario — the current 6 train episodes / 616 steps are too small), run the BC-warm-start-vs-RL-only benchmark comparison on SLURM/GPU, and produce checkpoint/training/evaluation manifests with durable pointers + the #1489 synthesis.

### Research contract

- Target claim/hypothesis: the BC loader + split/leakage gate + a BC training step execute end to end on the `expert_traj_v1` schema. **Engineering/smoke**, not a learning-quality hypothesis.
- Comparator: the prior fail-closed state where no BC loader or training step existed.
- Evidence tier: smoke (real-trajectory-smoke loading path).
- Result classification: loader/overfit smoke capability delivered and validated against a schema-faithful fixture; no benchmark/quality claim.
- Stop rule: stop at smoke; do not claim warm-start quality or any comparison result.
- Synthesis target: #1496 warm-start readiness and later #1489 synthesis only after qualified training/evaluation on a larger collection.

### Downstream propagation

#1496 is notified that the bounded loader/overfit smoke capability now exists and passes on CPU; the full training/evaluation acceptance criteria and public fail-closed readiness gates remain open and gated on the larger collection + SLURM/GPU compute.

This PR was produced by the Pi-harness/GLM-5.2 cheap implementation lane; the review gate hardens and merges.

Refs #1496

## Gate follow-ups

- #5938 — final readiness optional-lane timeout in `tests/benchmark/test_arm_isolation_parity.py`; infrastructure/environment gate blocker, not a PR semantic finding.

### CODE REVIEW

**Comment by ll7** [comment]
Cheap review+fix pass for observed head `0d33b5acc360c70f4b9d1b635760157d216fcd0a` (non-authoritative).

First-pass findings:
- Finding 1 — fixed: tests/training/test_oracle_imitation_bc_smoke.py:218 had dead code: `loaded = load_expert_trajectory_dataset` was assigned and never used, with a `# noqa: F841 (readability anchor)` suppressing the unused-variable warning. The symbol is imported (line 27) and used directly in 7 other test sites, so the local rebinding served no purpose and was misleading. Removed the line.
- Finding 2 — fixed: robot_sf/training/oracle_imitation_bc_smoke.py: `_flatten_observation` docstring claimed pedestrian fields are 'padded/truncated to a fixed slot count so every step shares a feature dimension'. The code does neither: `_flatten_observation` only ravels+concatenates per field, and `_stack_fixed_width` right-pads each vector to the batch's maximum feature width (no truncation, no fixed slot count). Rewrote the docstring to accurately describe the pad-to-batch-max behavior so a future reader does not assume pedestrian counts are bounded.

Local validation:
- `scripts/dev/run_worktree_shared_venv.sh --standalone -- uv run ruff check robot_sf/training/oracle_imitation_bc_smoke.py scripts/validation/run_oracle_imitation_bc_smoke.py tests/training/test_oracle_imitation_bc_smoke.py` — exit 0
- `scripts/dev/run_worktree_shared_venv.sh --standalone -- uv run ruff format --check robot_sf/training/oracle_imitation_bc_smoke.py tests/training/test_oracle_imitation_bc_smoke.py` — exit 0
- `DISPLAY= MPLBACKEND=Agg SDL_VIDEODRIVER=dummy scripts/dev/run_worktree_shared_venv.sh --no-freshness-check -- uv run pytest tests/training/test_oracle_imitation_bc_smoke.py -v` — exit 0
- `git diff --check` — exit 0

Pushed fix commit: `5f70ee6ed6c069c43ea078f207c3568800838ea2`.

The Codex gate retains the accept/merge decision.

**Comment by ll7** [comment]
## Gate findings — initial review at `c6e80ca9d2f228717035ac80539504e9d10021d0`

1. **HIGH — registry dataset integrity is optional.** `resolve_dataset_path()` only verifies `dataset_artifact.sha256` when it happens to be a string, while `validate_trace_uri_registry()` does not validate the separate `dataset_artifact` block. A malformed registry can therefore pass the trace gate and run an arbitrary private NPZ without a required 64-character checksum comparison. Make the dataset URI and checksum mandatory and fail closed on malformed registry YAML or dataset metadata.
2. **HIGH — split metadata does not cover all rows.** `validate_split_leakage_contract()` checks only episode IDs listed in metadata, but `flatten_observation_action_pairs()` trains from every NPZ row tagged `train`. An extra/unlisted row can therefore enter training without its seed participating in the leakage check. Require a one-to-one, duplicate-free match between metadata IDs, NPZ episode IDs, and split tags before training.
3. **MEDIUM — reported final loss is stale.** The training loop stores `loss.item()` before `optimizer.step()`, so `final_loss` and `loss_reduction` describe the loss before the last update rather than the final model state. Recompute the final loss after the loop under `no_grad()` before writing metrics/manifests.
4. **MEDIUM — required identity fields are coerced instead of validated.** `_episode_identity()` turns values such as `None` into the literal ID `"None"`, and `_episode_seeds()` accepts arbitrary values that `int()` can coerce. Reject missing/non-string episode IDs and non-integral, non-boolean seeds with `OracleImitationBcSmokeError` so malformed provenance cannot satisfy the split gate.

Review-thread enumeration: **0 open review threads**, including CodeRabbit. CodeRabbit’s rate-limit comment is advisory and is not an open review thread.

The PR body’s `Refs #1496` and explicit smoke-only claim boundary are appropriate; the remaining training/comparison criteria stay open on the parent issue.

gate-verdict: changes-requested @ c6e80ca9d2f228717035ac80539504e9d10021d0

**Comment by ll7** [comment]
## Gate repair update — `d645257c5bccc593b749870a90e906063e11c27b`

The four findings from the review at `c6e80ca9d2f228717035ac80539504e9d10021d0` are fixed in commit `d645257c5`:

1. **Fixed:** registry-backed smoke now requires a mapping, private dataset URI, concrete 64-character SHA-256, valid YAML, and an exact checksum match before loading the dataset.
2. **Fixed:** loader alignment checks cover every required episode-level array; split metadata must cover exactly the NPZ episode IDs, reject duplicates, and agree with every row tag.
3. **Fixed:** final loss is recomputed from the post-update model state under `no_grad()`.
4. **Fixed:** episode/scenario/split identities must be non-empty strings and seeds must be integer, non-boolean provenance values.

Validation on the repaired head:
- `python -m pytest tests/training/test_oracle_imitation_bc_smoke.py tests/training/test_oracle_imitation_warm_start_readiness.py tests/training/test_issue_1496_job_13520_registration.py -q` — **39 passed**.
- Ruff check/format and `git diff --check` — **passed**.

Review-thread enumeration: **0 open review threads**, including CodeRabbit. The rate-limit comment is advisory only.

GitHub CI for this repaired head is still running; the gate remains pending until the configured checks finish.

gate-verdict: changes-requested @ d645257c5bccc593b749870a90e906063e11c27b

**Comment by ll7** [comment]
## Gate verdict — accepted repaired head `d645257c5bccc593b749870a90e906063e11c27b`

1. **Fixed:** registry-backed smoke requires valid registry YAML, a private dataset URI, a concrete 64-character SHA-256, and an exact dataset checksum match.
2. **Fixed:** every required episode-level array is aligned; split metadata covers exactly the NPZ episode IDs, rejects duplicates, and agrees with every row tag.
3. **Fixed:** final loss is measured from the post-update model state.
4. **Fixed:** episode/scenario/split identities and integer seed provenance fail closed on malformed values.

Review-thread enumeration: **0 open review threads**, including CodeRabbit. CodeRabbit’s skipped/rate-limit status is advisory only.

Proof:
- All active GitHub checks for this exact head are green, including the body-contract rerun.
- Focused and adjacent local validation: **39 passed**; Ruff, formatting, and `git diff --check` passed.
- Final local readiness core lane: **7,449 passed, 19 skipped**. The unrelated optional-lane `test_arm_isolation_parity` timeout is tracked by [#5938](https://github.com/ll7/robot_sf_ll7/issues/5938), linked in the PR body, and is not waived prose-only.
- Closure is correct: this is the bounded smoke slice and `Refs #1496`; the parent issue’s larger training/comparison criteria remain open.

The dispatcher owns the subsequent merge; this gate did not merge or enable auto-merge.

gate-verdict: accepted @ d645257c5bccc593b749870a90e906063e11c27b
